awsshell/interaction.py

```diff
@@ -0,0 +1,120 @@
+import sys
+import jmespath
+
+from six import with_metaclass
+from abc import ABCMeta, abstractmethod
+
+from prompt_toolkit import prompt
+from prompt_toolkit.contrib.completers import WordCompleter
+
+
+class InteractionException(Exception):
+    """Base exception for errors in interactions."""
+
+
+class Interaction(with_metaclass(ABCMeta, object)):
+    """Base Interaction class.
+
+    The base interaction class only has one constraint: an execute method that
+    will take some data as input, apply arbirtary logic on that data to perform
+    a transformation and then return the resulting data to be used in later
+    stages.
+    """
+
+    def __init__(self, model, prompt):
+        self._model = model
+        self.prompt = prompt
+
+    @abstractmethod
+    def execute(self, data):
+        """Execute the interaction, transforming the data in some way."""
+
+
+class SimpleSelect(Interaction):
+    """Displays a list of options, allowing the user to select one.
+
+    Given a list of one or more items, display them in a dropdown selection
+    menu and allows the user to pick one. If a path is present on the
+    interaction the path will be applied to each item and the result will be
+    used as the string to display in the menu. Upon selection, the related item
+    is returned. If no path is present, the list is assumed to be of str and
+    used as is.
+    """
+
+    def __init__(self, model, prompt_msg, prompter=prompt):
+        super(SimpleSelect, self).__init__(model, prompt_msg)
+        self._prompter = prompter
+
+    def execute(self, data):
+        if not isinstance(data, list) or len(data) < 1:
+            raise InteractionException('SimpleSelect expects a non-empty list')
+        if self._model.get('Path') is not None:
+            disp_data = jmespath.search(self._model['Path'], data)
+            cmpltr = WordCompleter(disp_data, sentence=True, ignore_case=True)
```

Comment:
I did this to get the line under 79 characters. I went back and forth on this but I think I can agree it's more readable as the whole word. I'll refactor this so it's broken up on multiple lines.